### PR TITLE
Sort app scripts topologically by its dependencies

### LINKED ISSUE #30278: Issues with dependency logic in `Util::addScript`
The new dependency logic as implemented in https://github.com/nextcloud/server/pull/30015 is broken if dependencies themselves have dependencies. E.g. when text app depends on viewer and viewer depends on files, the script from text is loaded first while the script from viewer comes later:

The problem is that the `files` dependency of `viewer` moves `viewer/js/viewer-main` to the `last` element of `files`, while the `viewer` dependendy on `text` moves `text/js/text-viewer` to the `last` element of `viewer`. Viewer is before files but `viewer` no longer contains `viewer/js/viewer-main`, so loading `text/js/text-viewer` from `last` fails.

`self::$scripts` from https://github.com/nextcloud/server/blob/791d57051ad4d5fd6f24a0e8c200ca1fc8555c3e/lib/public/Util.php#L243 when loading the Collectives app:

```json
{
  "accessibility": {
    "first": [
      "accessibility/l10n/en",
      "accessibility/js/accessibilityoca"
    ]
  },
  "files_sharing": {
    "first": [
      "files_sharing/l10n/en",
      "files_sharing/js/dist/main"
    ]
  },
  "text": {
    "first": [
      "text/l10n/en"
    ]
  },
  "viewer": {
    "first": [
      "viewer/l10n/en"
    ],
    "last": [
      "text/js/text-viewer"
    ]
  },
  "files": {
    "first": [],
    "last": [
      "viewer/js/viewer-main"
    ]
  },
  "theming": {
    "first": [
      "theming/l10n/en",
      "theming/js/theming"
    ]
  },
  "user_status": {
    "first": [
      "user_status/l10n/en",
      "user_status/js/user-status-menu"
    ]
  },
  "core": {
    "first": [],
    "last": [
      "core/js/dist/unified-search"
    ]
  }
}
```

After removing `files` dependency from `viewer` (in https://github.com/nextcloud/viewer/pull/1081/files#diff-a14aae69340720190ddf327b54675eacd97116b45f5b57f583ddb52829a3f152R41):

```json
{
  "accessibility": {
    "first": [
      "accessibility/l10n/en",
      "accessibility/js/accessibilityoca"
    ]
  },
  "files_sharing": {
    "first": [
      "files_sharing/l10n/en",
      "files_sharing/js/dist/main"
    ]
  },
  "text": {
    "first": [
      "text/l10n/en"
    ]
  },
  "viewer": {
    "first": [
      "viewer/l10n/en",
      "viewer/js/viewer-main"
    ],
    "last": [
      "text/js/text-viewer"
    ]
  },
  "theming": {
    "first": [
      "theming/l10n/en",
      "theming/js/theming"
    ]
  },
  "user_status": {
    "first": [
      "user_status/l10n/en",
      "user_status/js/user-status-menu"
    ]
  },
  "core": {
    "first": [],
    "last": [
      "core/js/dist/unified-search"
    ]
  }
}
```

So I guess we need to improve the logic and sort scripts by dependencies.

### PR DESCRIPTION
**Pull request by mejo-**
Implement a proper topological sorting algorithm. Based on the
implementation by https://github.com/marcj/topsort.php

Throws a CircularDependencyException if a circular dependency is
detected.

Fixes: #30278

### CODE REVIEW

**Comment by mejo-** [comment]
Not sure whether we want the CircularDependencyException to be thrown. We also could skip that if we don't consider circular dependencies a problem (in certain cases they can result in broken dependency chains).

**Comment by mejo-** [comment]
> Not sure whether we want the CircularDependencyException to be thrown. We also could skip that if we don't consider circular dependencies a problem (in certain cases they can result in broken dependency chains).

Thinking about it again, it's not a good idea to fail hard here by throwing an exception and we should just log an error instead. Agreed?

**Comment by mejo-** [comment]
> Thinking about it again, it's not a good idea to fail hard here by throwing an exception and we should just log an error instead. Agreed?

I removed the CircularDependencyException. Instead, the circular dependency is now merely logged as an error if detected.

**Review by skjnldsv — DISMISSED** [review]

File: lib/public/Util.php [hunk 1]
```diff
@@ -198,46 +203,86 @@ public static function addScript($application, $file = null, $afterAppId = null)
 			self::addTranslations($application);
 		}
 
-		// store dependency if defined
-		if (!empty($afterAppId)) {
-			self::$scriptDeps[$application] = $afterAppId;
+		// store app in dependency list
+		if (!array_key_exists($application, self::$scriptDeps)) {
+			self::$scriptDeps[$application] = (object)[
```
Comment (ChristophWurst) [inline lib/public/Util.php:208]: :-1: We don't use objects this way. Preferably create your own little POPO with getters/setters and use that as your typed data structure or use an untyped associative array.

File: lib/public/Util.php [hunk 2]
```diff
@@ -198,46 +203,86 @@ public static function addScript($application, $file = null, $afterAppId = null)
...
 	}
 
+	/**
+	 * Recursive topological sorting
+	 *
+	 * @param object $app
+	 * @param array|null $parents
+	 * @since 24.0.0
+	 */
+	private static function topSortVisit(object $app, array &$parents = null): void {
```
Comment (ChristophWurst) [inline lib/public/Util.php:227]: This probably comes down to my personal preferences, but let's try our best to avoid any statics. Static variables and methods should be used as rare as possible. This algorithm could be moved into a sideeffect free helper class that works in a functional way: you pass in an array of scripts and their dependencies and as a result you get the sorted scripts. Then use this functionality in here. The algorithm itself will be 100% unit testable. The ugliness of static APIs can be hidden in here until we find a better way.

In summary: static API to add a script can stay. Internals of the algorithm can be improved.

Your future self, coworkers, static analysis and your IDE will thank you :)

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: lib/public/Util.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/public/Util.php:227]: On another thought: the server and this area is not my direct responsibility. If others are okay with this code then I'll revoke my remarks :v:

Comment (mejo-) [reply lib/public/Util.php:227]: Mh, I'm happy to implement your suggestions. But not sure whether I get the picture. I thought that all functions in Util.php were static for performance reasons. You suggest to outsource the whole topologicial sorting into a separated class - but its functions would need to be static as well in order to use them in the static function `getScripts()`, no?

Comment (ChristophWurst) [reply lib/public/Util.php:227]: You can take the registered scripts from the static variable, create an instance of the algorithm class, pipe the data through it and return the result.

File: lib/public/Util.php (hunk 1 above)
Comment (mejo-) [reply lib/public/Util.php]: Agreed. You mean like this?: https://github.com/nextcloud/server/pull/30385/commits/9ff203cceeabd41efbd8f6b21adb383ea003877b

Comment (ChristophWurst) [reply lib/public/Util.php]: Yes :)

File: lib/public/Util.php (hunk 2 above)
Comment (mejo-) [reply lib/public/Util.php]: Did this now: https://github.com/nextcloud/server/pull/30385/commits/0b5adc805a61732ba49a3aff74c4ac4c53a5d22e Is that how you meant it?

File: lib/public/Util.php (hunk 1 above)
Comment (mejo-) [reply lib/public/Util.php]: Yay :blush:

File: lib/public/Util.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/public/Util.php]: Looks good! Move the inputs from the constructor args to the method, the `sortedScriptDeps` to a local variable of `sort` and it's perfect :)

You get bonus points if you inject a `LoggerInterface` via the constructor and use that instead of ``\OCP\Util::writeLog`` :wink:

Comment (mejo-) [reply lib/public/Util.php]: > Looks good! Move the inputs from the constructor args to the method, the `sortedScriptDeps` to a local variable of `sort` and it's perfect :)

Thanks @ChristophWurst :blush:

Done the first now in https://github.com/nextcloud/server/pull/30385/commits/cb763035f63a582dac4372c8195665de883a5617. `$this->sortedScriptDeps` cannot be local to `sort()` as it's updated by `topSortVisit()`, no?
 
> You get bonus points if you inject a `LoggerInterface` via the constructor and use that instead of `\OCP\Util::writeLog` wink

Done in https://github.com/nextcloud/server/pull/30385/commits/86a90fe10930779242c76d48b1ec989f81c4cb82

File: lib/private/AppScriptSort.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	private $logger;
+
+	/** @var AppScriptDependency[] */
+	private $scriptDeps;
+
+	/** @var string[] */
+	private $sortedScriptDeps;
+
+	public function __construct() {
+		$this->logger = \OC::$server->get(LoggerInterface::class);
```
Comment (skjnldsv) [inline lib/private/AppScriptSort.php:43]: Not possible via DI ?

**Review by skjnldsv — APPROVED** [review]
This took a crazy turn! Code looks good, tests too! :rocket: 
Awesome work!

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
Psalm failure unrelated

File: lib/public/Util.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/public/Util.php]: You could pass it through `topSortVisit` by reference. Same for `scriptDeps`.

File: lib/private/AppScriptSort.php (hunk 3 above)
Comment (mejo-) [reply lib/private/AppScriptSort.php:43]: I wouldn't know how - at least not if used by static class/function `Util::getScripts()`. Or maybe there's a way and I simply don't know about it? :thinking:

File: lib/public/Util.php (hunk 2 above)
Comment (mejo-) [reply lib/public/Util.php]: What would be the advantage of passing `sortedScriptDeps` on and on through all iterations of `topSortVisit()` over making it a global variable to the class?

Comment (ChristophWurst) [reply lib/public/Util.php]: It will be free of side effects

File: lib/private/AppScriptSort.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/private/AppScriptSort.php:43]: add it as regular constructor arg and in Util you use `\OC::$server->get(AppScriptSort::class)` instead of `new AppScriptSort()`. DI will resolve the argument and build the sort algorithm instance for you.

Comment (mejo-) [reply lib/private/AppScriptSort.php]: Uh nice, thanks for the explanation :heart_eyes: Implemented now in b770347537b8a516fed855f04f32485d1fe6cda8. It also has the advantage that we can check for `$logger->error()` being called in the unit test, which allows to test the circular dependency detection.

File: lib/public/Util.php (hunk 2 above)
Comment (mejo-) [reply lib/public/Util.php]: I see. Not sure whether I fully get why this is relevant here, but implemented as you requested nevertheless :wink: See 4a94d0109641a66e52d6aad5856391cd8cec2594

**Review by ChristophWurst — APPROVED** [review]
:+1: :shipit:

**Comment by mejo-** [comment]
> This took a crazy turn!

Soooo true :see_no_evil: :zany_face: 

I wouldn't have expected this to become such a ride. Thanks a lot to @ChristophWurst and @skjnldsv for guiding me!

**Comment by mejo-** [comment]
Seems like I'm not allowed to merge - maybe due to the (unrelated) psalm test failures? :thinking: Could someone with super powers merge? :grimacing:

File: lib/public/Util.php [hunk 4]
```diff
@@ -45,6 +46,9 @@
 
 namespace OCP;
 
+use OC\AppScriptDependency;
```
Comment (CarlSchwan) [inline lib/public/Util.php:49]: Unfortunately, this causes an psalm warning, since now OCP depends on OC. This wasn't found earlier since psalm was broken but it appear on https://github.com/nextcloud/server/pull/30508

I'm not sure if we can ignore this issue and just add it to the baseline or if we should fix it

Comment (juliushaertl) [reply lib/public/Util.php:49]: Fine to add to the baseline from my perspective as it is only used in the method implementation and not exposed in any kinds.
